packages/apollo-server-plugin-base/src/index.ts

```diff
@@ -15,9 +15,9 @@ export {
   WithRequired,
 };
 
-export interface ApolloServerPlugin {
+export interface ApolloServerPlugin<TContext = Record<string, any>> {
   serverWillStart?(service: GraphQLServiceContext): ValueOrPromise<void>;
-  requestDidStart?<TContext>(
+  requestDidStart?(
```

Comment:
I do believe that implementing the `TContext` at the `interface` level of `ApolloServerPlugin`, rather than on `requestDidStart`, is the correct direction here.

However, while I'm unconvinced that the prior incarnation of the `TContext` type argument to `requestDidStart` (here) was **ever** useful, I do believe that someone could have implemented it.

I wonder if there's a way to default to using the newly added `TContext` above but keep accepting whatever existing users by, for example:

```suggestion
  requestDidStart?<TContext = TContext>(
```

That said, since this is a compile time error, maybe we should just commit this potentially breaking change that fixes what we otherwise believe to be completely broken behavior and see if any issues are raised from it?  It would be a compile time error, so I'm comfortable with that, knowing that it won't be a production-surprise for anyone.  If someone is affected, they can open an issue with their use case and we can try to accommodate it.

Thoughts?

Reply:
I was unable to provide any context type to this `requestDidStart`, it still treated my function as a generic one, not allowing to narrow, so it's clearly a bug and should be fixed - it's just one change, so after upgrade users might just update the signature to the one that exist and doesn't have `any` 😄

Reply:
I'd consider this change or any change in the spirit of this unnecessary. In my last example from the [playground reproduction](https://bit.ly/2qyEKc2), we see that a user can still provide their own type argument in the same exact manner as before with the new implementation.

Reply:
Ok. Agree.  Resolving this comment unactioned.